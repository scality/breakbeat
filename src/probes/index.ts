import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { NoopProbe, noopProbeSchema } from './NoopProbe';
import { KafkaConsumerLagProbe, kafkaConsumerLagProbeSchema } from './KafkaConsumerLagProbe';
import { Probe } from './Probe';

const probeSchema = joi.alternatives().try(
    noopProbeSchema.required(),
    kafkaConsumerLagProbeSchema.required(),
);

export type ProbeConfig = joi.extractType<typeof probeSchema>;

function buildProbe(probeConfig: ProbeConfig): Probe {
    const type = probeConfig.type;
    switch (type) {
    case 'noop':
        return new NoopProbe(probeConfig);
    case 'kafkaConsumerLag':
        return new KafkaConsumerLagProbe(probeConfig);
    }

    // not really dead code for transpiled uses
    const tpe: string = probeConfig['type'];
    throw new Error(`unsupported probe type '${tpe}'`);
}

export {
    Probe,
    buildProbe,
    probeSchema,
};
