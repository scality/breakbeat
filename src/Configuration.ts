import * as joi from '@hapi/joi';
import 'joi-extract-type';

import { probeSchema } from './probes';

export const defaultProbeEvaluateInterval = 60000;

const topSchema = joi.object({
    probes: joi.array().items(probeSchema).optional(),
    nominalEvaluateIntervalMs: joi.number().default(defaultProbeEvaluateInterval).optional(),
    trippedEvaluateIntervalMs: joi.number().default(defaultProbeEvaluateInterval).optional(),
    stabilizingEvaluateIntervalMs: joi.number().default(defaultProbeEvaluateInterval).optional(),
    stabilizeAfterNSuccesses: joi.number().default(2).optional(),
});

export type Config = joi.extractType<typeof topSchema>;

export function validate(config: unknown): Config {
    if (!config) {
        return validate({});
    }

    const res = topSchema.validate(config);
    if (res.error) {
        throw res.error;
    }

    return res.value;
}

export const configWithDefaults = validate({});
