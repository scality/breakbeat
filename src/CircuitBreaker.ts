import { Config, defaultProbeEvaluateInterval, validate } from './Configuration';
import { Probe, buildProbe } from './probes';

export enum BreakerState {
    Nominal = 0,
    Stabilizing = 1,
    Tripped = 2,
}

export class CircuitBreaker {
    _config: Config;
    _probes: Probe[];

    _evaluateIntervalByAggregateState: {
        [key in BreakerState]: number;
    };

    _aggregateState: BreakerState;
    _stabilizingCounter: number;
    _stabilizeThreshold: number;
    _evaluateTimeoutHandle: NodeJS.Timeout | null;

    // a promise that unit tests can await on, as jest's fake timers
    // get confused with async flows
    _evaluatingPromiseHook: Promise<unknown> | null;

    constructor(config: unknown) {
        this._config = validate(config);
        this._probes = (this._config.probes || []).map(buildProbe);

        this._aggregateState = BreakerState.Nominal;
        this._stabilizingCounter = 0;

        this._stabilizeThreshold = this._config.stabilizeAfterNSuccesses || 0;
        this._evaluateIntervalByAggregateState = {
            [BreakerState.Nominal]: this._config.nominalEvaluateIntervalMs || defaultProbeEvaluateInterval,
            [BreakerState.Stabilizing]: this._config.stabilizingEvaluateIntervalMs || defaultProbeEvaluateInterval,
            [BreakerState.Tripped]: this._config.trippedEvaluateIntervalMs || defaultProbeEvaluateInterval,
        };

        this._evaluateTimeoutHandle = null;
        this._evaluatingPromiseHook = null;
    }

    get state(): BreakerState {
        return this._aggregateState;
    }

    start() {
        this._aggregateState = BreakerState.Nominal;
        this._scheduleNextEvaluation();
    }

    stop() {
        if (this._evaluateTimeoutHandle) {
            clearTimeout(this._evaluateTimeoutHandle);
            this._evaluateTimeoutHandle = null;
        }
    }

    async _evaluate() {
        this._evaluatingPromiseHook = Promise.allSettled(this._probes.map(p => p.check()));
        await this._evaluatingPromiseHook;
        this._evaluatingPromiseHook = null;

        const allOk = this._probes.every(probe => probe.value);

        if (allOk) {
            switch (this._aggregateState) {
            case BreakerState.Tripped:
                this._aggregateState = BreakerState.Stabilizing;
                this._stabilizingCounter = 1;
                break;

            case BreakerState.Stabilizing:
                this._stabilizingCounter++;

                if (this._stabilizingCounter >= this._stabilizeThreshold) {
                    this._aggregateState = BreakerState.Nominal;
                }
                break;

            default:
                break;
            }
        } else {
            this._aggregateState = BreakerState.Tripped;
        }

        // only schedule next if start was called before
        if (this._evaluateTimeoutHandle) {
            this._scheduleNextEvaluation();
        }
    }

    _scheduleNextEvaluation() {
        this._evaluateTimeoutHandle = setTimeout(
            () => void this._evaluate(),
            this._evaluateIntervalMs,
        );
    }

    get _evaluateIntervalMs(): number {
        return this._evaluateIntervalByAggregateState[this._aggregateState];
    }
}
