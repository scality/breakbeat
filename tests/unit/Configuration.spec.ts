import { describe, expect, test } from '@jest/globals';

import { validate, configWithDefaults } from '../../src/Configuration';

describe('Configuration', () => {
    test('should validate', () => {
        const config = {};
        expect(validate(config)).toStrictEqual(configWithDefaults);
    });

    test('should accept missing configuration', () => {
        expect(() => validate(undefined)).not.toThrow();
    });

    test('should reject invalid configuration', () => {
        const config = {
            probes: 2,
        };
        expect(() => validate(config)).toThrowErrorMatchingSnapshot();
    });
});
