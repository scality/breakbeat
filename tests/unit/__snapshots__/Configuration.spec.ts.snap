// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration should reject invalid configuration 1`] = `""probes" must be an array"`;
